### PR TITLE
[lldb][dwarf] Compute fully qualified names on simplified template names with DWARFTypePrinter

### PR DESCRIPTION
**Pull request by ZequanWu**
This is the second half of https://github.com/llvm/llvm-project/pull/90008.

Essentially, it replaces the work of resolving template types when we just need the qualified names with walking the DIE tree using `DWARFTypePrinter`.

### Result
For an internal target, the time spent on `expr *this` for the first time reduced from 28 secs to 17 secs.

### CODE REVIEW

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFBaseDIE.h [hunk 1]
```diff
@@ -46,6 +48,7 @@ class DWARFBaseDIE {
   explicit operator bool() const { return IsValid(); }
 
   bool IsValid() const { return m_cu && m_die; }
+  bool isValid() const { return IsValid(); }
```
Comment (labath) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFBaseDIE.h:51]: How about using `operator bool` as the common api ?

**Review by labath — COMMENTED** [review]
I'd say this looks very good. Given this is a relatively novel approach, I'd suggest we  give folks a bit more time to take a look at this.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFBaseDIE.h [hunk 2]
```diff
@@ -95,6 +100,8 @@ class DWARFBaseDIE {
 
   const char *GetName() const;
 
+  const char *getShortName() const { return GetName(); }
```
Comment (Michael137) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFBaseDIE.h:103]: Nit: How do people feel about grouping all the new APIs that are required for LLVM compatbility together and commenting on adding a comment stating why we have them?

File: lldb/test/Shell/SymbolFile/DWARF/x86/simplified-template-names.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+// Test that we following DW_AT_signature correctly. If not, lldb might confuse the types of v1 and v2.
+// RUN: %clangxx --target=x86_64-pc-linux -g -gsimple-template-names -fdebug-types-section %s -o %t
+// RUN: %lldb %t -o "target variable v1 v2" -o exit | FileCheck %s --check-prefix=TYPE
+
+// LOG: unique name: ::t2<outer_struct1::t1<int> >
+// LOG: unique name: ::t2<outer_struct2::t1<int> >
+
+// TYPE:      (t2<outer_struct1::t1<int> >) v1 = {}
+// TYPE-NEXT: (t2<outer_struct2::t1<int> >) v2 = {}
```
Comment (Michael137) [inline lldb/test/Shell/SymbolFile/DWARF/x86/simplified-template-names.cpp:18]: Do we need to check the typename that gets displayed here? Is that actually affected by this patch?

File: lldb/test/Shell/SymbolFile/DWARF/x86/simplified-template-names.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+// Test against logging to see if we print the fully qualified names correctly.
+// RUN: %clangxx --target=x86_64-pc-linux -g -gsimple-template-names %s -o %t
+// RUN: %lldb %t -o "log enable dwarf comp" -o "target variable v1 v2" -o exit | FileCheck %s --check-prefix=LOG
+
+// Test that we following DW_AT_signature correctly. If not, lldb might confuse the types of v1 and v2.
+// RUN: %clangxx --target=x86_64-pc-linux -g -gsimple-template-names -fdebug-types-section %s -o %t
+// RUN: %lldb %t -o "target variable v1 v2" -o exit | FileCheck %s --check-prefix=TYPE
+
+// LOG: unique name: ::t2<outer_struct1::t1<int> >
+// LOG: unique name: ::t2<outer_struct2::t1<int> >
```
Comment (Michael137) [inline lldb/test/Shell/SymbolFile/DWARF/x86/simplified-template-names.cpp:15]: Do we strictly need this test? Technically the existing `-gstimple-template-names` test should already test this codepath right?

A more interesting test could be a unit-test that uses the `DWARFTypePrinter` with LLDB DWARFDIEs. Not sure how difficult that is to set up though.

File: lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp [hunk 5]
```diff
@@ -1403,26 +1403,6 @@ static TemplateParameterList *CreateTemplateParameterList(
...
-    llvm::ArrayRef<TemplateArgument> pack_args =
-        template_param_infos.GetParameterPackArgs();
-    args.append(pack_args.begin(), pack_args.end());
-  }
-  std::string str;
-  llvm::raw_string_ostream os(str);
-  clang::printTemplateArgumentList(os, args, GetTypePrintingPolicy(),
-                                   template_param_list);
-  return str;
-}
```
Comment (Michael137) [inline lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:1424]: Nice

**Review by Michael137 — COMMENTED** [review]
Overall LGTM, just left some nits

I think separating out the "DWARFDIE compatbility" changes from the "use DWARFTypePrinter for type lookup" would make it easier to review, but I don't mind too much

File: lldb/test/Shell/SymbolFile/DWARF/x86/simplified-template-names.cpp (hunk 4 above)
Comment (labath) [reply lldb/test/Shell/SymbolFile/DWARF/x86/simplified-template-names.cpp:15]: I'm not exactly thrilled by the idea of using logs for testing, and even less about adding log statements with the sole purpose of testing a piece of code.

A unit test for the DWARFTypePrinter would definitely be better, and I think it should be possible to reuse what we have in unittests/SymbolFile/DWARF/DWARFDIETest.cpp. For testing its integration into the rest of lldb, an end-to-end might be better (maybe the existing ones are fine, but we can definitely add new ones if you think something is missing).

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFBaseDIE.h (hunk 2 above)
Comment (labath) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFBaseDIE.h]: (I was considering the same thing, but I don't particularly care either way)

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFBaseDIE.h (hunk 1 above)
Comment (labath) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFBaseDIE.h]: I'm sorry, I guess I wasn't clear about this. What I was suggesting is to just revert the addition `isValid` -- as there already exists a comon API. I didn't want to remove things that are already there. (I mean, I'd be fine with that doing that under the flag of moving the LLDB API closer to LLVM, but I don't think we ought to do that as a part of this PR)

**Review by labath — APPROVED** [review]

**Comment by rastogishubham** [comment]
This patch broke the lldb incremental bot on greendragon

https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/15486/

******************** TEST 'lldb-shell :: SymbolFile/DWARF/x86/simplified-template-names.cpp' FAILED ********************
Exit Code: 1

Command Output (stderr):
--
RUN: at line 7: /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang --driver-mode=g++ --target=specify-a-target-or-use-a-_host-substitution --target=x86_64-pc-linux -g -gsimple-template-names /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/Shell/SymbolFile/DWARF/x86/simplified-template-names.cpp -o /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/tools/lldb/test/Shell/SymbolFile/DWARF/x86/Output/simplified-template-names.cpp.tmp
+ /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang --driver-mode=g++ --target=specify-a-target-or-use-a-_host-substitution --target=x86_64-pc-linux -g -gsimple-template-names /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/Shell/SymbolFile/DWARF/x86/simplified-template-names.cpp -o /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/tools/lldb/test/Shell/SymbolFile/DWARF/x86/Output/simplified-template-names.cpp.tmp
ld: warning: -m is obsolete
ld: unknown option: --hash-style=gnu
clang: error: linker command failed with exit code 1 (use -v to see invocation)

**Comment by rastogishubham** [comment]
I reverted this change with c51786b022bbff6c902cbcd2af5cc6535f6d9b5d to keep the bot green

**Comment by ZequanWu** [comment]
> This patch broke the lldb incremental bot on greendragon
> 
> https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/15486/
> 
> ******************** TEST 'lldb-shell :: SymbolFile/DWARF/x86/simplified-template-names.cpp' FAILED ******************** Exit Code: 1
> 
> ## Command Output (stderr):
> RUN: at line 7: /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang --driver-mode=g++ --target=specify-a-target-or-use-a-_host-substitution --target=x86_64-pc-linux -g -gsimple-template-names /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/Shell/SymbolFile/DWARF/x86/simplified-template-names.cpp -o /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/tools/lldb/test/Shell/SymbolFile/DWARF/x86/Output/simplified-template-names.cpp.tmp
> 
> * /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang --driver-mode=g++ --target=specify-a-target-or-use-a-_host-substitution --target=x86_64-pc-linux -g -gsimple-template-names /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/Shell/SymbolFile/DWARF/x86/simplified-template-names.cpp -o /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/tools/lldb/test/Shell/SymbolFile/DWARF/x86/Output/simplified-template-names.cpp.tmp
>   ld: warning: -m is obsolete
>   ld: unknown option: --hash-style=gnu
>   clang: error: linker command failed with exit code 1 (use -v to see invocation)

I should explicitly call ld.lld. Relanded [here](https://github.com/llvm/llvm-project/commit/f06c187799d910fd3ac3e9106397e5eecff9f265)

**Pull request merged**